---
title: "Tutorial 01: Single-Cell Transcriptomics"
author: "Dr. Sonny Chen & Keon Arbabi"
date: "26/06/2021"
output: 
  html_document: 
    keep_md: yes
---

### **About `.Rmd` files**

This is an R Markdown notebook, where chunks of code can be executed independently and the output is visible immediately beneath the input. Accordingly, related lines of code can be organized and executed together. 

To do this click the green "play" icon at the top right of the chunk, or Ctrl + Shift + Enter (macOS: Cmd + Shift + Enter) to run the current chunk. An indicator will appear on the left margin to show you execution progress. Lines of code that have been sent to R are marked with dark green; lines that have not yet been sent to R are marked with light green.

The current working directory inside a notebook chunk is always the directory containing the notebook .Rmd file. You’ll get a warning if you try to change the working directory inside a notebook chunk, and the directory will revert back to the notebook’s directory once the chunk is finished executing.

Lastly, Markdown notebooks can easily be converted or "knit" into formats such as HTML, PDF, and Word for reproducible sharing of R code and figures. 

If you wish to learn more visit: https://bookdown.org/yihui/rmarkdown/

### **About this tutorial**

**Main learning objective:** become familiar with single-cell RNA sequencing data and how to analyze its features using Seurat.  

Furthermore, you will:

  * Be introduced to the concept of "reads" and "counts"
  * Import and normalize a counts matrix
  * Perform dimensionality reduction 
  * Visualize clusters of cells
  * Visualize gene expression across cells
  * Identify cell type-specific marker genes (i.e. differential expressed genes, DEGs)
  
### **Load packages**

We will begin by loading the packages necessary for this tutorial, which are already installed in your docker container. If the package is not loading, uncomment the line for installing it. 

`Tidyverse` is a collection of R packages that share a simple syntax for data structures    
`Seurat` is an R toolkit for single cell genomics. (Satija et al., 2015; Hao et al., 2021; https://satijalab.org/seurat/index.html)   
`MAST` is the statistical approach we will use to find DEGs (Finak et al., 2015)    

```{r, load packages, inlcude=FALSE, message=FALSE, warning=FALSE}

# install.packages("tidyverse")
library(tidyverse)

# install.packages("Seurat") # install Seurat, as needed 
library(Seurat) 

# install.packages("BiocManager") # MAST is a bioconductor package, install BiocManager as needed
# BiocManager::install("MAST") # install MAST, as needed
library(MAST)

# install.packages("knitr")
library(knitr) # for markdown formatting

```

### **Load data**

The dataset we're working with includes 893 human cells from multiple cortical areas collected by the Allen Institute for Brain Science (AIBS) using scRNA-seq. Learn more here: https://portal.brain-map.org/explore/classes

First, let's get the metadata for our cells.

```{r, load metadata}

# get metadata for each cell
# includes sample id, brain region, cortical layer, and cell-type labels curated by the AIBS
meta_data = read.csv(file = "./data/AIBS_scRNAseq_2019_human_metadata_downsampled.csv")
# set transcriptomic sample id as the rownames, this is important for Seurat 
rownames(meta_data) = meta_data$sample_name
# preview 
kable(meta_data[1:5,1:12], row.names = F)

```

Next we'll import our gene expression data. 

RNA-seq generally begins with the isolation and purification of RNA. RNA fragments are converted to cDNA, which are flanked by specific "adapter" sequences and amplified by PCR. These steps are referred to as "library preparation." 

The cDNA library is then sequenced. Each raw *read* is the sequence of nucleotide bases of a cDNA molecule. 

Reads generated by RNA-seq experiments must ultimately be aligned, or "mapped" to a reference genome. Read alignment to a reference provides biological information in two basic ways: 

  * First, it generates a dictionary of the genomic features represented in each sample. Aligned reads become annotated, and the highly fragmented sampling of RNA sequences is connected to the gene families, individual transcripts, small RNAs, or individual exons expressed in the original tissue sample or cell.

  * Second, the number of reads aligned to each feature approximates abundances of those features in the original sample. This is the *counts matrix* that we import here. Each column represents an individual cell and each row represents a gene. The integers in each cell is a *count* of the reads that map onto each gene per sample. 


```{r, load count matrix}

# get counts matrix
counts_matrix = read.csv(file = "./data/AIBS_scRNAseq_2019_human_matrix_downsampled_genefiltered.csv", 
                         row.names = 1, # set the gene names in the first column to rownames 
                         check.names = FALSE) # prevent automatic renaming of columns)
# preview 
kable(counts_matrix[10:18,1:8])

```

### **Create Seurat object and normalize**

We can now create our Seurat object, which stores our counts matrix, metadata, and the results of any normalization step or analyses.

```{r, create Seurat object}

# create our Seurat object
seu_obj  = CreateSeuratObject(count = counts_matrix, 
                              meta.data = meta_data) 
# the number of genes (features) and cells (samples)
# matches the corresponding dimensions of our initial data
seu_obj
# note where our data is stored 
kable(seu_obj@meta.data[1:5,1:8])
kable(seu_obj@assays$RNA@counts[15:20,1:8])

```

We next need to normalize our count data. This is important to consider as, for example, we do not want to confound increased gene expression because of biological reasons with increased RNA amplification due to technical differences. 

By default, we employ a global-scaling normalization method “LogNormalize” that normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor, and log-transforms the result. 

```{r, normalize data}

# normalize counts data in seurat object
# we use a scale factor of 1 million to convert our values to counts-per-million
seu_obj = NormalizeData(seu_obj, 
                        normalization.method = "LogNormalize", 
                        scale.factor = 1000000)
# preview normalized data
kable(seu_obj@assays$RNA@data[15:20,1:8])

```

### **Dimensionality reduction and clustering**

Our Seurat object already contains metadata labels assigning each cell to a cell type subclass curated by the AIBS. Some steps involved in classifying cells include dimensionality reduction and clustering.

First, we apply a linear transformation (‘scaling’) that is a standard pre-processing step prior to dimensional reduction techniques like Principal Component Analysis (PCA). This shifts the expression of each gene, so that the mean expression across cells is 0 and the variance across cells is 1. This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate. 

PCA is the process of identifying a series of orthogonal vectors that iteratively capture the greatest amount of variance in the data. The aim of this approach is to adequately summarize the high-dimensional relationships of the original data in a lower-dimensional space. 

We perform PCA on the 2000 most variable features then order both cells and features according to their PCA scores for visualization.

```{r, PCA, fig.width=16, fig.height=9}
# scale for all genes
seu_obj = ScaleData(seu_obj,
                    features = rownames(seu_obj))
# find most variable features
seu_obj = FindVariableFeatures(seu_obj,
                               nfeatures = 2000)
# run PCA
seu_obj = RunPCA(seu_obj,
                 verbose = FALSE)
# preview principal components (PCs)
print(seu_obj@reductions$pca, dims = 1:5, nfeatures = 5)
# visualize cells and features for first two PCs
DimHeatmap(seu_obj, dims = 1:2, cells = 300, nfeatures = 50, balanced = TRUE)

```

Next well use a graph-based clustering approach to group similar cells. 

`FindNeighbors` computes a K-nearest neighbor (KNN) graph based on euclidean distances in PCA space, calculated previously. "Edges" (straight lines) are drawn between two cells with similar feature expression patterns. Edge weights are then refined and the graph is partitioned into highly interconnected "communities." 

`FindClusters` uses modularity optimization techniques such as the Louvain algorithm, which are outside the scope of this tutorial. 

```{r, find clusters and visualize, fig.width=12, fig.height=7, inlcude=FALSE, message=FALSE, warning=FALSE}

# create KNN graph of data
seu_obj = FindNeighbors(seu_obj, verbose = F) 
# find clusters using default parameters 
seu_obj = FindClusters(seu_obj, verbose = F)
# run umap for a number of dimensions (from PCA), we'll go with 10
seu_obj = RunUMAP(seu_obj, dims = 1:10, verbose = F) 

# Visualize clusters 
# we use umap as our reduction method; we could put PCA here, but it would be harder to interpret 
DimPlot(seu_obj,
        reduction = "umap", 
        label = TRUE) + # show the labels of our clusters 
        NoLegend()

```

The AIBS has provided cell type classification in their metadata. We can visualize these groups on our umap plot. 

```{r, show AIBS cell types on UMAP, fig.width=12, fig.height=7}

# The Idents() function lets us specify which grouping variable we want to work with
Idents(seu_obj) =  "subclass_label" 

### run the same code to generate our UMAP plot, but now we have different active "identities" due to the code above
DimPlot(seu_obj, 
        reduction = "umap", 
        label = TRUE) 
```

### **Find cell-type markers**

We are ready to find cell-type markers. There are a few different options available to us depending on our needs. However, all methods require us to first specify how we are grouping cells. For example, we might want to group cells by cell class (such as inhibitory neurons) or subclass (such as SST cells). Or maybe we want to compare gene expression between cells from different brain regions. To specify this, we set the active "identities" of cells in our Seurat object to a metadata variable/column within the Seurat object.

```{r, set identity}

# set active identity
Idents(seu_obj) = "subclass_label" 
# see the cell types we'll be contrasting to find markers 
unique(seu_obj$subclass_label)

```

Seurat can help you find markers that define cell subtypes via differential expression between them.

```{r, find markers for all groups, fig.width=18, fig.height=9, inlcude=FALSE, message=FALSE, warning=FALSE}

# find markers for all cell subtypes 
all_markers = FindAllMarkers(object = seu_obj,
                             logfc.threshold = 2.5, # only test genes with >2.5x log-fold change between groups compared 
                             min.pct = .35, # only test genes detected over this fraction of cells for the group of interest
                             only.pos = T, # only return positive markers (enriched in the cell group of interest)
                             test.use = "MAST", # the statistical test to use
                             verbose = F) %>% 
  rename_at("cluster", ~ "subclass_label") # rename column, necessary for merge()
# preview of our results
kable(all_markers[1:10,]) 

# expression heatmap of cells and features
# get top 5 markers for each subclass label
top_genes = all_markers %>%
  group_by(subclass_label) %>%
  top_n(n = 10, wt = avg_log2FC)
# plot
DoHeatmap(seu_obj, features = top_genes$gene, size = 4) + 
  NoLegend()

```

Save marker genes for later. These can be used for many applications. 

```{r, save markers}

# keep track of the class-level classifications (e.g. Glutamatergic) for later 
# merge class_labels from metadata by matching to subclass labels 
all_markers = merge(all_markers, distinct(meta_data, class_label, subclass_label, .keep_all = F), by = "subclass_label")
# rearrange columns 
all_markers = all_markers %>% 
  dplyr::select(gene, subclass_label, class_label, everything())

# save
write.csv(all_markers, file = "./data/human_markers.csv")

```

We are using some sample threshold values for logfc.threshold and min.pct here. These are values chosen based on our previous work and experience. These values may be considered somewhat high, meaning that fewer genes will pass the threshold for testing and that less time will be needed to run `FindAllMarkers`. In a more serious analysis, we would adjust these thresholds based on multiple factors.

If we are not interested in all groups, but rather are interested in only one or a few subclasses, we can use the `FindMarkers` function instead.

```{r, find markers for Olgiodendrocytes, fig.width=16, fig.height=9, inlcude=FALSE, message=FALSE, warning=FALSE}

# Using the oligodendrocyte subclass as an example
olgio_markers = FindMarkers(object = seu_obj, 
                            ident.1 = "Oligodendrocyte", # specify our group of interest here
                            logfc.threshold = 2.5, 
                            min.pct = .35, 
                            only.pos = T,
                            test.use = "MAST",
                            verbose = F) %>%
# sort by high to low average log-fold change
  arrange(desc(avg_log2FC))
# preview our results
head(olgio_markers)

# we can use violin plots to verify that our top oligodendrocyte markers are indeed enriched in our oligodendrocyte cluster (green)
VlnPlot(seu_obj, 
        features = rownames(olgio_markers)[1:3])
# similarly, we can visualize marker expression on our umap plot
FeaturePlot(seu_obj, 
            features = rownames(olgio_markers)[1:4])


```

Note the arguments are largely the same between `FindMarkers` and `FindAllMarkers`. The code above compares oligodendrocyte cells with a group containing all other cells in the dataset. If we wanted to compare between cells of two specific groups defined in `subclass_label`, we can use the "ident.2" argument in `FindMarkers`.

```{r, find markers between two groups, inlcude=FALSE, message=FALSE, warning=FALSE}
# For example, we want to find positive markers in SST cells but only compared to VIP cells
SST_vs_VIP_markers = FindMarkers(object = seu_obj, 
                                 ident.1 = "SST", 
                                 ident.2 = "VIP", # setting the specific "other" group to compare  
                                 logfc.threshold = 2.5, 
                                 min.pct = .35, 
                                 only.pos = T,
                                 test.use = "MAST",
                                 verbose = F) 
head(SST_vs_VIP_markers) # preview our results

```

This concludes the tutorial on Seurat. 

Sources: 

1. https://satijalab.org/seurat/index.html
2. https://github.com/RGLab/MAST
3. https://portal.brain-map.org/atlases-and-data/rnaseq/human-multiple-cortical-areas-smart-seq
4. https://rnaseq.uoregon.edu/#exp-design 



